File: codechef/march17/pshtbrth.py
#!/bin/python3
allmask = 1 << 16
grund = [0] * allmask
recmask = []
for i in range(4):
    for j in range(4):
        for x in range(i+1, 5):
            for y in range(j+1, 5):
                now = 0
                for p in range(i, x):
                    for q in range(j, y):
                        now ^= (1 << (p * 4 + q))
                recmask.append(now)
                
for mask in range(allmask):
    st = set()
    for v in recmask:
        if v & mask == v:
            st.add(grund[mask ^ v])
    grund[mask] = 0
    ls = list(st)
    ls.sort()
    for v in ls:
        grund[mask] += grund[mask] == v

bit = [0] * (5 * 10 ** 5 + 10)
n = 0
def upd(x, v):
    global bit, n
    while x <= n:
        bit[x] ^= v
        x += x & (-x)

def get(x):
    global bit, n
    ret = 0
    while x > 0:
        ret ^= bit[x]
        x -= x & (-x)
    return ret

for _ in range(int(input())):
    n, m = map(int, input().split())
    a = []
    for i in range(n+1):
        bit[i] = 0
    for i in range(n):
        mask = 0
        for j in range(4):
            s = input()
            for k in range(4):
                if s[k] == '1':
                    mask |= (1 << (j * 4 + k))
        input()
        a.append(grund[mask])
        upd(i+1, a[i])
        
    for __ in range(m):
        ty, *q = map(int, input().split())
        if ty == 1:
            ans = get(q[0]-1) ^ get(q[1])
            print('Pishty' if ans else 'Lotsy')
        else:
            mask = 0
            for j in range(4):
                s = input()
                for k in range(4):
                    if s[k] == '1':
                        mask |= (1 << (j * 4 + k))
            upd(q[0], grund[mask] ^ a[q[0]-1])
            a[q[0]-1] = grund[mask]

File: codechef/march17/pshtbrth.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 5e5 + 5;
char s[50];
int grund[(1 << 16) + 10];
int a[N];
int bit[N], n;
void upd(int x, int v) {
  for (; x <= n; x+=x&-x) bit[x] ^= v;
}
int get(int x) {
  int ret = 0;
  for (; x; x-=x&(-x)) ret ^= bit[x];
  return ret;
}

int main() {
  vector< int > recmask;
  for (int i = 0; i < 4; i++) {
    for (int j = i+1; j < 5; j++) {
      for (int x = 0; x < 4; x++) {
        for (int y = x+1; y < 5; y++) {
          int now = 0;
          for (int u = i; u < j; u++) {
            for (int v = x; v  < y; v++) {
              now |= (1<< (u * 4 + v));
            }
          }
          recmask.push_back(now);
        }
      }
    }
  }
  int allmask = 1<<16;
  for (int mask = 0; mask < allmask; mask++) {
    set< int > st;
    st.clear();
    for (int v : recmask)
      if ((v & mask) == v) {
        assert((mask ^ v) <  mask);
        st.insert(grund[mask ^ v]);
      }
    grund[mask] = 0;
    for (int v : st) {
      grund[mask] += (grund[mask] == v);
    }
  }

  int t;
  scanf("%d", &t);
  while (t--) {
    int m;
    scanf("%d %d", &n, &m);
    for (int i = 0; i <= n; i++)
      bit[i] = 0;
    for (int i = 0; i < n; i++) {
      int mask = 0;
      for (int x= 0; x < 4; x++) {
        scanf("%s", s);
        for (int y = 0; y < 4; y++)
          if (s[y] == '1')
            mask |= (1<<(x * 4 + y));
      }
      a[i] = grund[mask];
      upd(i+1, a[i]);
    }
    while (m--) {
      int ty;
      scanf("%d", &ty);
      if (ty == 1) {
        int l, r;
        scanf("%d %d", &l, &r);
        int ans = get(l-1) ^ get(r);
        puts(ans ? "Pishty" : "Lotsy");
      }
      else {
        int pos, mask = 0;
        scanf("%d", &pos);
        for (int i = 0; i < 4; i++) {
          scanf("%s", s);
          for (int j = 0; j < 4; j++)
            if (s[j] == '1')
              mask |= (1<< (i * 4 + j));
        }
        upd(pos, a[pos-1] ^ grund[mask]);
        a[pos-1] = grund[mask];
      }
    }
  }
  return 0;
}